File: v1/src/main/java/com/api/v1/people/services/PersonUpdateServiceImpl.java
package com.api.v1.people.services;

import org.springframework.stereotype.Service;

import com.api.v1.people.domain.PersonAuditTrail;
import com.api.v1.people.domain.PersonAuditRepository;
import com.api.v1.people.domain.PersonCrudRepository;
import com.api.v1.people.Person;
import com.api.v1.people.PersonUpdateDTO;
import com.api.v1.people.PersonUpdateService;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;

@Service
public class PersonUpdateServiceImpl implements PersonUpdateService {

    private final PersonCrudRepository repository;
    private final PersonAuditRepository auditTrailRepository;

    public PersonUpdateServiceImpl(PersonCrudRepository repository, PersonAuditRepository auditTrailRepository) {
        this.repository = repository;
        this.auditTrailRepository = auditTrailRepository;
    }

    @Override
    public Person update(@NotNull Person person, @Valid PersonUpdateDTO updatingDto) {
        PersonAuditTrail auditTrail = PersonAuditTrail.of(person);
        PersonAuditTrail savedAuditTrail = auditTrailRepository.save(auditTrail);
        Person updatedPerson = person.update(updatingDto);
        return repository.save(updatedPerson);
    }
}

File: v2/src/main/kotlin/com/api/v2/people/services/PersonUpdateServiceImpl.kt
package com.api.v2.people.services

import com.api.v2.people.domain.PersonAuditTrail
import com.api.v2.people.domain.PersonAuditRepository
import com.api.v2.people.domain.PersonCrudRepository
import com.api.v2.people.Person
import com.api.v2.people.PersonUpdateDTO
import com.api.v2.people.PersonUpdateService
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.springframework.stereotype.Service

@Service
class PersonUpdateServiceImpl(
    val auditTrailRepository: PersonAuditRepository,
    val repository: PersonCrudRepository
) : PersonUpdateService {

    override fun update(
        person: @NotNull Person,
        updatingDTO: @Valid PersonUpdateDTO
    ): Person {
        val auditTrail = PersonAuditTrail.of(person)
        val savedAuditTrail = auditTrailRepository.save(auditTrail)
        val updatedPerson = person.update(updatingDTO)
        return repository.save(updatedPerson)
    }
}